### PR TITLE
Properly truncate referenced sources

### LINKED ISSUE #4741: Referenced source is truncated too much
For a source line of
```
    function newAnnouncement(announcementType Type, string Announcement, string Link, bool Oppositable, string _str, uint256 _uint, address _addr) onlyOwner external {
```
the source reference in the error is truncated to
```
publisher.sol:119:105: Error: Data location must be "calldata" for parameter in external function, but none was given.
 ... string _str ... 
     ^---------^
```

It should show way more context.

The relevant file is `SourceReferenceFormatter.cpp`

### PR DESCRIPTION
**Pull request by iamonuwa**
### Your checklist for this pull request

Please review the [guidelines for contributing](http://solidity.readthedocs.io/en/latest/contributing.html) to this repository.

Please also note that this project is released with a [Contributor Code of Conduct](CONDUCT.md). By participating in this project you agree to abide by its terms.

### Checklist
- [x] Code compiles correctly
- [x] All tests are passing
- [ ] New tests have been created which fail without the change (if possible)
- [ ] README / documentation was extended, if necessary
- [ ] Changelog entry (if change is visible to the user)
- [x] Used meaningful commit messages

### Description
Increase the length of the error message to accommodate more context about the error.
Thank you for your help!

Fixes #4741

### CODE REVIEW

File: libsolidity/interface/SourceReferenceFormatter.cpp [hunk 1]
```diff
@@ -55,7 +55,7 @@ void SourceReferenceFormatter::printSourceLocation(SourceLocation const* _locati
 		}
 		if (line.length() > 150)
 		{
-			line = " ... " + line.substr(startColumn, locationLength) + " ... ";
+			line = " ... " + line.substr(0, 75) + " ... ";
```
Comment (chriseth) [inline libsolidity/interface/SourceReferenceFormatter.cpp]: Hm, I think this should rather be something like `"..." + line.substr(max(0, startColumn - 35), min(locationLength + 70, line.length())) + " ....";`, plus removing the "..." if nothing was actually cut off.

**Review by chriseth — CHANGES_REQUESTED** [review]
It's a good start, but I think it still needs to be adjusted a little.

Could you please also add a test in `./test/cmdlineTests.sh`?

**Comment by iamonuwa** [comment]
Yeah, sure. Not really good with c++ tests

**Comment by chriseth** [comment]
Could you please remove the `...` if nothing was removed like I mentioned in https://github.com/ethereum/solidity/pull/4801#discussion_r209551492 ? Thanks!

Please also add a test for your changes.

**Comment by iamonuwa** [comment]
@chriseth any guide to writing tests for this please?

**Comment by chriseth** [comment]
Take a look at `cmdlinetests.sh`, especially those tests that use `test_solc_file_input_failures`.

**Comment by iamonuwa** [comment]
@chriseth alright, what about checking if there is no text remaining on truncate?

**Comment by vs77bb** [comment]
@chriseth I think @iamonuwa might be awaiting your feedback here!

**Comment by chriseth** [comment]
@vs77bb ah thanks for mentioning, it somehow slipped my inbox.

@iamonuwa I'm sorry, but I don't really understand what you mean. Can you provide an example, please?

**Comment by iamonuwa** [comment]
@chriseth I've figured it out. Writing test should submit this task tonight.

**Comment by liangdzou** [comment]
@iamonuwa Since you are back for this issue. Shall I merge my code to your branch? Then, you can take care of the rest. I have added 8 test cases. Also, I think the implementation is ok.

**Comment by liangdzou** [comment]
@vs77bb @iamonuwa @chriseth 
FYI, I am not interested in the bounty. @iamonuwa  can have all of them no matter which PR is merged finally.

**Comment by iamonuwa** [comment]
@liangdzou alright.

**Comment by ekpyron** [comment]
Superseded by #4941

**Pull request closed without merging**